### PR TITLE
MINIFI-236: Make GetFile, PutFile, TailFile, and ExecuteProcess ThreadSafe

### PR DESCRIPTION
**Pull request by phrocker**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by apiri** [comment]
reviewing

File: CMakeLists.txt [hunk 1]
```diff
@@ -136,6 +136,21 @@ enable_testing(test)
     target_include_directories(tests PRIVATE BEFORE "libminifi/include/provenance")
     target_link_libraries(tests ${CMAKE_THREAD_LIBS_INIT} ${UUID_LIBRARIES} ${LEVELDB_LIBRARIES} ${OPENSSL_LIBRARIES} minifi yaml-cpp c-library civetweb-cpp)
     add_test(NAME LibMinifiTests COMMAND tests)
+
```
Comment (phrocker) [inline CMakeLists.txt]: Can we try and make this less duplicative?

File: libminifi/include/core/ProcessSession.h [hunk 2]
```diff
@@ -122,6 +128,7 @@ class ProcessSession {
   void penalize(std::shared_ptr<core::FlowFile> &flow);
   void penalize(std::shared_ptr<core::FlowFile> &&flow);
 // Import the existed file into the flow
+  void importFrom(io::DataStream &stream, std::shared_ptr<core::FlowFile> &&flow);
```
Comment (phrocker) [inline libminifi/include/core/ProcessSession.h]: better comment?

File: libminifi/include/processors/GetFile.h [hunk 3]
```diff
@@ -79,16 +80,27 @@ class GetFile : public core::Processor {
...
+   */
+  void onSchedule(
+        core::ProcessContext *context,
+        core::ProcessSessionFactory *sessionFactory);
   // Initialize, over write by NiFi GetFile
   virtual void initialize(void);
   // perform directory listing
-  void performListing(std::string dir);
+  void performListing(std::string dir,const GetFileRequest &request);
+
```
Comment (phrocker) [inline libminifi/include/processors/GetFile.h]: comment

File: libminifi/include/utils/ThreadPool.h [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+  Worker<T>& operator = (Worker<T>&&) ;
+
+  std::shared_ptr<std::promise<T>> getPromise();
+
+private:
+   std::function<T()> task;
+   std::shared_ptr<std::promise<T>> promise;
+};
+
+template< typename T>
```
Comment (phrocker) [inline libminifi/include/utils/ThreadPool.h]: comment to remind people that && will take ownership of Worker

File: libminifi/src/core/ProcessSession.cpp [hunk 5]
```diff
@@ -530,6 +530,91 @@ void ProcessSession::read(std::shared_ptr<core::FlowFile> &&flow,
   }
 }
 
+void ProcessSession::importFrom(io::DataStream &stream,
+                            std::shared_ptr<core::FlowFile> &&flow) {
+  std::shared_ptr<ResourceClaim> claim = std::make_shared<ResourceClaim>();
+
+  int max_read = getpagesize();
```
Comment (phrocker) [inline libminifi/src/core/ProcessSession.cpp]: is this portable?

File: libminifi/src/processors/TailFile.cpp [hunk 6]
```diff
@@ -145,10 +145,10 @@ static bool sortTailMatchedFileItem(TailMatchedFileItem i,
                                     TailMatchedFileItem j) {
   return (i.modifiedTime < j.modifiedTime);
 }
-void TailFile::checkRollOver() {
+void TailFile::checkRollOver(std::string fileLocation, std::string fileName) {
```
Comment (phrocker) [inline libminifi/src/processors/TailFile.cpp]: const arguments?

File: libminifi/test/TestExecuteProcess.cpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+  std::atomic<bool> is_ready(false);
+
+  std::vector<std::thread> processor_workers;
+
+  core::ProcessorNode node2(processor);
+  std::shared_ptr<core::ProcessContext> contextset = std::make_shared<
+      core::ProcessContext>(node2, test_repo);
+  core::ProcessSessionFactory factory(contextset.get());
+ // processor->onSchedule(contextset.get(), &factory);
```
Comment (phrocker) [inline libminifi/test/TestExecuteProcess.cpp]: uncomment.

Comment (phrocker) [inline libminifi/test/TestExecuteProcess.cpp]: unless this should be removed.

File: libminifi/include/processors/GetFile.h [hunk 8]
```diff
@@ -101,23 +113,17 @@ class GetFile : public core::Processor {
...
-  int64_t _maxAge;
-  int64_t _minSize;
-  int64_t _maxSize;
-  bool _ignoreHiddenFile;
-  int64_t _pollInterval;
-  int64_t _batchSize;
-  uint64_t _lastDirectoryListingTime;
-  std::string _fileFilter;
+
+  std::map<std::string,std::atomic<uint64_t>> last_listing_times_;
```
Comment (apiri) [inline libminifi/include/processors/GetFile.h]: Not sure I follow the usage of the map.  Presumably we only have one directory per instance as there is no input.

**Comment by apiri** [comment]
@phrocker Just wanted to confirm this was done before a final review.  I think it likely makes sense to get this one in before we merge #73

**Comment by apiri** [comment]
reviewed the latest commit and code looks good

Verified builds in OS X 10.11, Centos 7.x, and OS X 10.12 (via #75)

Will get this merged and thank you for the fixes.

**Pull request closed without merging**